core/controllers/async_download.py

```diff
@@ -0,0 +1,147 @@
+
+    # Upload the file to S3 and get presigned URL
+    file_obj = io.BytesIO(response.data)
+    bucket = Config.AWS_S3_BUCKET_FIND_DATA_FILES
+    find_service_base_url = Config.FIND_SERVICE_BASE_URL
+    email_template_id = Config.NOTIFY_FIND_REPORT_DOWNLOAD_TEMPLATE_ID
+
+    if not bucket:
+        current_app.logger.error("AWS_S3_BUCKET_FIND_DATA_FILES is not set.")
+    elif not find_service_base_url:
+        current_app.logger.error("FIND_SERVICE_BASE_URL is not set.")
+    elif not email_template_id:
+        current_app.logger.error("NOTIFY_TEMPLATE_ID is not set.")
+    else:
+        current_datetime = datetime.now().strftime("%Y-%m-%d-%H:%M:%S")
+        file_name = f"fund-monitoring-data-{current_datetime}.{file_format}"
+        url = upload_file(file=file_obj, bucket=bucket, object_name=file_name, expiry=86400)
+
+        # email the URL to the user
+        send_email(
+            template_id=email_template_id,
+            email_address=email_address,
+            download_url=url,
+            find_service_url=find_service_base_url + "/download",
+            file_format=file_format_str,
+            file_size_str=file_size_str,
+        )
+
+
+def upload_file(file: IO | FileStorage, bucket: str, object_name: str, expiry: int) -> str:
+    """Upload a file to an S3 bucket, generate presigned url and return it
+    :param file: File to upload,
+    :param bucket: Bucket to upload to,
+    :param object_name: S3 object name (filename),
+    :param expiry: URL expiry time in seconds,
+
+    :return: Presigned URL
+    """
+
+    _S3_CLIENT.put_object(Bucket=bucket, Key=object_name, Body=file)
+
+    url = _S3_CLIENT.generate_presigned_url(
+        "get_object",
+        Params={"Bucket": bucket, "Key": object_name},
+        ExpiresIn=expiry,
+    )
+
+    return url
+
+
+def send_email(
+    template_id: str, email_address: str, download_url: str, find_service_url: str, file_format: str, file_size_str: str
+):
+    """Send an email to the user with the download link.
+    :param template_id: NOTIFY email template ID,
+    :param email_address: user's email address,
+    :param download_url: download URL,
+    :param find_service_url: URL of the FIND service,
+    :param file_format: human-readable file format string,
+    :param file_size_str: human-readable file size,
+    """
+
+    try:
+        notifications_client = NotificationsAPIClient(current_app.config["NOTIFY_FIND_API_KEY"])
+    except KeyError:
+        current_app.logger.error("NOTIFY_FIND_API_KEY is not set.")
+    else:
+        notifications_client.send_email_notification(
+            email_address=email_address,
+            template_id=template_id,
+            personalisation={
+                "download_url": download_url,
+                "find_service_url": find_service_url,
+                "file_format": file_format,
+                "file_size": file_size_str,
+            },
+        )
```

Comment:
Can we call this something more specific than `send_email`? I think we can either inline this function, or make it more specific. Maybe something like `send_email_for_find_download`?

Also, this function takes a `template_id` but it's not generic for all templates because we're passing specific personalisation. If we're keeping this function, then maybe just put the template ID directly in the function?

Reply:
Function renamed, template_id removed from function parameters and template ID has been put directly in the function.